packages/core/src/ContextManager.ts

```diff
@@ -118,12 +119,15 @@ export class ContextManager {
             const lastPredicate = predicates.pop();
             let currentLastPredicatePosition = -1;
             if (lastPredicate) {
-                const found = [...currentAncestors]
-                    .reverse()
-                    .find(node => node.test(lastPredicate));
+                const found = [...currentAncestors].find(node => node.test(lastPredicate));
                 if (found) {
-                    currentLastPredicatePosition = currentAncestors.indexOf(found);
-                    currentAncestors.splice(currentLastPredicatePosition);
+                    const index = currentAncestors.indexOf(found);
+                    currentLastPredicatePosition = currentAncestors.length - 1 - index;
+                    // todo: decide this line or the previous line is clearer
+                    // currentLastPredicatePosition = [...currentAncestors].reverse().indexOf(found);
+                    // remove all ancestors so we do not check them again later
```

Comment:
I suggest "// Remove already tested ancestors to avoid retesting them later on."